storage/innobase/fil/fil0crypt.cc

```diff
@@ -836,6 +836,65 @@ static inline void fil_crypt_read_crypt_data(fil_space_t *space) {
   mtr.commit();
 }
 
+/**
+Write crypt data belonging to space to page0
+@param space tablespace with crypt data to write
+*/
+static void fil_crypt_write_crypt_data_to_page0(fil_space_t *space) {
+  mtr_t mtr;
+  mtr.start();
+
+  if (buf_block_t *block = buf_page_get_gen(
+          page_id_t(space->id, 0), page_size_t(space->flags), RW_X_LATCH, NULL,
+          Page_fetch::NORMAL, __FILE__, __LINE__, &mtr)) {
+    space->crypt_data->write_page0(
+        space, block->frame, &mtr, space->crypt_data->min_key_version,
+        space->crypt_data->type, space->crypt_data->encryption_rotation);
+  }
+  mtr.commit();
+}
+
+struct MutexGuard {
+  MutexGuard(ib_mutex_t *mutex) : mutex(mutex) { mutex_enter(mutex); };
+  ~MutexGuard() { mutex_exit(mutex); }
+
+ private:
+  ib_mutex_t *mutex;
+};
+
+bool fil_crypt_exclude_tablespace_from_rotation(fil_space_t *space) {
+  MutexGuard fil_crypt_threads_mutex_guard(&fil_crypt_threads_mutex);
+
+  if (space->crypt_data == nullptr) {
+    space->crypt_data = fil_space_create_crypt_data(FIL_ENCRYPTION_OFF,
```

Comment:
In all places crypt_data is first assigned to space before calling write_page0. There is even an assert in write_page0: ut_ad(this == space->crypt_data);. If we want to change that maybe we should do that as a separate bug and let this PR follow the current behavior ?

Reply:
IMHO this is simple request. Again separate PR, scheduling is un-necessary overhead.

all I ask is to do is
```
fil_space_crypt_t* crypt_data = fil_space_create_crypt_data()
fil_crypt_write_crypt_data_to_page0(crypt_data);
space->crypt_data = crypt_data
```
Also you claim it is race-free, so only doing this at place should be fine right?

In my analysis (below), space->crypt_data is read/written under different mutexes at different times. So it is really hard for me..

1. fil_space_set_crypt_data under  mutex_enter(&crypt_data->mutex);
2. fil_crypt_read_crypt_data() uses some kind of double checking & space->size to avoid race. No latch protection
3. some readers under  mutex_enter(&fil_crypt_threads_mutex);
4. fil_crypt_rotate_page() reads space->crypt_data without any protection
5. fil_space_crypt_close_tablespace(sp); without any protection. Releases shard mutex as well :(
6. fil_space_crypt_get_status() used by I_S , no protection. Uses double check of space->crypt_data poniter
7. may other usages that I am yet to analyze..

Reply:
I do not want to do it as part of this PR as crypt_data is assigned first to space and then written to page0 in many places. It needs correction in all those places, not just here.

Reply:
ok, fine.

Reply:
PS-6107